### PR TITLE
Widget Example and Documentation Suite

### LINKED ISSUE #747: Switching examples to dojo cli
**Enhancement** 

With the creation of the dojo build tool at https://github.com/dojo/cli we should be using this for our widget examples so that:
- There are less steps involved to host the application (currently the developer needs to run `grunt dev`, `serve` and `tsc -w` if they want it watched). `dojo build` will take care of all these in one
- Easier updates in the future
- More consistency with different dojo projects.
- Easier dev url, currently it's `http://localhost:5000/_build/common/example/`. It would optimal for this to be  `http://localhost:5000/`

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Introduces a new widget example and documentation application that can be used to view changes during development and also generates additional documentation such as the properties interface, theme css classnames and the readme contents for each widget when built in "docs" mode.

This includes only a basic example for each widget type to demonstrate how to build examples, the existing examples will need to be ported over from their current location.

Remaining Tasks: 

 * [ ] Add missing jsdoc for the widget properties
 * [ ] Add landing page for all the widgets
 * [ ] Add link to go back to the landing page
 * [ ] Port over the existing examples into the new example format
 * [ ] Deploy documentation on merge into master

Example Deployment: https://dist-jcj0dvpig.now.sh/

Resolves #747

### CODE REVIEW

File: src/examples/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+To run units tests in node only use `npm run test` or `dojo test` which uses JIT (just in time) compilation.
+
+To run the unit tests against built bundles, run `npm run test:unit`.
+
+To be re-run the unit tests without needing to re-build the full application each time, first build the app with `--mode unit` and the `--watch` option, `dojo build --mode unit --watch`. Then run `dojo test --config local` to run the unit tests as needed.
+
+The build test artifacts are written to the `output/tests/unit` directory. These tests are located in the `tests/unit` directory.
+
+## Running functional tests
```
Comment (samenz) [inline src/examples/README.md]: Tom got rid of the all the functional tests for widgets so I think this can be taken out

Comment (agubler) [reply src/examples/README.md]: That's part of the auto generated README by cli-create-app (which I used to start the examples application). It's not the README for the actual @dojo/widgets repo.

File: src/examples/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+import * as css from './App.m.css';
+
+const widgetFilenames = getWidgetFileNames(configs);
+
+interface AppProperties {
+	includeDocs: boolean;
+}
+
+const factory = create({ block }).properties<AppProperties>();
```
Comment (samenz) [inline src/examples/src/App.tsx]: What's this block middleware for?

File: src/examples/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+							return null;
+						}
+						const widgetPath = `${widgetName}/${example.filename}`;
+						const content =
+							widgetExampleContent[`${widgetPath}.tsx`] ||
+							widgetExampleContent[`${widgetPath}.ts`];
+						const propertyInterface = widgetProperties[widgetName];
+						const themeClasses = widgetThemeClasses[widgetName];
+						return (
+							<virtual key={widgetPath}>
```
Comment (samenz) [inline src/examples/src/App.tsx]: What's `<virtual>` doing here? Never seen this before.

File: src/examples/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+								<SideMenu name={widgetName} config={widgetConfig} />
+								<div classes={[css.content]}>
+									{isBasic && includeDocs && <div innerHTML={readmeContent} />}
+									<h1>{isBasic ? 'Basic Usage' : example.title}</h1>
+									<Example content={content}>
+										<example.module />
```
Comment (samenz) [inline src/examples/src/App.tsx]: Interesting way to create a dom element, never thought to do this before 👍

File: src/examples/src/App.tsx (hunk 3 above)
Comment (agubler) [reply src/examples/src/App.tsx]: `<virtual>` is a special node that can be used when you want to return an array in tsx, there is more information here https://dojo.io/learn/creating-widgets/rendering-widgets#tsx-widget-example

File: src/examples/src/ActiveLink.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+	diffProperty('to', (current: ActiveLinkProperties, next: ActiveLinkProperties) => {
+		if (current.to !== next.to) {
+			const router = injector.get<Router>(routerKey);
+			const currentHandle = cache.get<Handle>('handle');
+			if (currentHandle) {
+				currentHandle.destroy();
+			}
+			if (router) {
+				const handle = router.on('outlet', ({ outlet }) => {
```
Comment (samenz) [inline src/examples/src/ActiveLink.ts]: I'm guessing this validates that the path `to` points to is a valid path? Seems like this code block is here twice, maybe make a function for it?

Comment (agubler) [reply src/examples/src/ActiveLink.ts]: If I abstract it to a function, then I'll lose the type inference and have to import a type which I don't really want to do.

File: src/examples/src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import block from '@dojo/framework/core/middleware/block';
```
Comment (samenz) [inline src/examples/src/App.tsx]: What does this block middleware do?

Comment (agubler) [reply src/examples/src/App.tsx]: There is some information here in the docs https://dojo.io/learn/building/buildtime-rendering#dojo-blocks

**Comment by samenz** [comment]
I pulled down this branch to take a look at what the new examples page looks like. It seems when I try to set `includeDocs` to true I get alot of `'n' implicitly has an 'any' type` build errors. Seems like you may need to add types to these files.

Additionally would you want to add a toggle of some sort to turn the readme documentation on/off?

**Comment by samenz** [comment]
Two more things.
- Should the old example under `src/common/examples` be taken out?
- We may want to add an additional issue to remove the css build warnings when this gets merged in.

Thanks for doing this btw! This is going to make developing dojo widgets infinitely easier.

**Comment by agubler** [comment]
@samenz you should be able to pass `-f docs=true` on the command line to run the examples with docs.

As for the errors they should be resolved with a reinstall of the node_modules (with the latest)

File: package.json [hunk 7]
```diff
@@ -17,7 +17,7 @@
     "url": "https://github.com/dojo/widgets.git"
   },
   "scripts": {
-    "examples": "serve",
+    "examples": "dojo build app -m dev -w -s",
```
Comment (tomdye) [inline package.json]:
```suggestion
    "dev": "dojo build app -m dev -w -s",
```

File: README.md [hunk 8]
```diff
@@ -207,26 +206,61 @@ To test against browsers with a local selenium server run:
...
+}
+```
 
-### Viewing widget examples locally
+ * filename: The name of the widget module, defaults to `index`
+ * overview: The configuration for the basic example including the imported Basic module and the example filename (has to be `'Basic'`)
+ * examples: Additional examples for the widget, an array of configuration that specifies the title, description, module and example filename. 
 
-Each Dojo widget includes functioning example code so you can view the widget. To view individual widget example:
+ To view the examples locally run `npm run examples` and navigate to http://localhost:9999, this starts the examples in watch mode and should update widget module are changed.
```
Comment (tomdye) [inline README.md]:
```suggestion
 To view the examples locally run `npm run dev` and navigate to http://localhost:9999, this starts the examples in watch mode and should update widget module are changed.
```

**Pull request merged**